### ISSUE TITLE
Optimise performance of the Patterns Toolkit API endpoint and add missing features

### ISSUE THREAD

**Issue by ramonjd** [Automattic/wp-calypso#45946]
D49914-code added in the ptk/patterns API endpoint.

In Phase 2 of the Patterns Toolkit (PTK) project (pbAok1-1uA-p2) we want to:

- improve the endpoint (features and performance) in collaboration with the View team.
- help investigate performant ways of adding Patterns to the editor.

Ideas / requests and to do list includes:

* [x] Review https://github.com/Automattic/wp-calypso/pull/45926 and add any missing features required to help get it shipped.
* [x] Add object cache to the translations at the endpoint to speed up non English pattern fetching, along with a query param to invalidate / refresh the cache (D50178-code — note, this was rolled back as it was no longer needed)
* [x] Improve performance of the API endpoint by optimising when we switch locales (D50471-code, and pMz3w-c2X-p2 and pbxlJb-qk-p2#comment-537)
* [ ] Add support for custom viewport values (raised in pbxlJb-oL-p2#comment-498).
* [ ] Add support for cache invalidation of consumers of the endpoint (this will require a bit of exploration to come up with the right solution), raised in https://github.com/Automattic/wp-calypso/pull/45926#discussion_r495613952. Or an alternate approach for handling invalidation.
* [ ] Add `wp_locale` query param that accepts a WP locale string and maps it to the correct language code for use in wpcom. This might be essentially moving the mapping from (for example) `pt_br` to `pt-br` to the endpoint, instead of consuming code needing to implement it (more discussion from https://github.com/Automattic/wp-calypso/pull/45926/files#r494712732).
* [x] Add `type` query param to filter by a 'hidden' category of `pattern` or `layout` with that category hidden from the categories array of each returned translated pattern. The idea is to cheat a kind of "pattern type" feature that is separate from the visible categories on the pattern, that will then map to pattern categories when the pattern is registered in the editor. (raised in https://github.com/Automattic/wp-calypso/pull/45926#discussion_r496088909) — there's a possible alternative of using hierarchical categories instead, suggested here: https://github.com/Automattic/wp-calypso/pull/45926#discussion_r496480219 — update: this to do list item is marked as complete, but we didn't need to add the `type` feature after all, instead we're using tags for this kind of grouping and the `tags` param in the endpoint for filtering.
* [ ] Write up a feature request / idea or proposal for registering patterns via JS instead of PHP
* [x] Revisit caching at the endpoint to guard against frequent calls (p58i-9yt-p2#comment-47647, D50987-code, and pMz3w-c6M-p2)
* [x] Add `pattern_meta` custom taxonomy and key to API responses (D53825-code and pbxlJb-zn-p2)

**Comment by andrewserong** [comment]
Ideas / requests so far include:

* [ ] Add object cache to the translations at the endpoint to speed up non English pattern fetching, along with a query param to invalidate / refresh the cache.
* [ ] Add support for custom viewport values (raised in pbxlJb-oL-p2#comment-498).
* [ ] Add support for cache invalidation of consumers of the endpoint (this will require a bit of exploration to come up with the right solution), raised in https://github.com/Automattic/wp-calypso/pull/45926#discussion_r495613952.
* [ ] Add `wp_locale` query param that accepts a WP locale string and maps it to the correct language code for use in wpcom. This might be essentially moving the mapping from (for example) `pt_br` to `pt-br` to the endpoint, instead of consuming code needing to implement it (more discussion from https://github.com/Automattic/wp-calypso/pull/45926/files#r494712732).

**Comment by andrewserong** [comment]
Most of the intended enhancements from this issue are now complete. For the outstanding issues:

* Add support for custom viewport values — this has been made easier by the addition pattern_meta (PCYsg-uUe-p2). There's already an open issue for this (77-gh-Automattic/view-design) so it doesn't need to be tracked in this issue.
* Add support for cache invalidation of consumers of the endpoint — this will be easier to do once D55247-code lands as it refactors the cache key / pattern fetching code in a way where it'll be easier to invalidate. We haven't had a need to do this yet, so I would not worry about cache invalidation and deal with it if and when the problem comes up.
* Add wp_locale query param that accepts a WP locale string and maps it to the correct language code for use in wpcom — this is still a neat idea, however after this issue was opened, the `get_iso_639_locale` function in the ETK plugin was refactored to be defined in a single location, so it's less of a pain: https://github.com/automattic/wp-calypso/blob/5930d248ee81f1df5566e368668da9ae6c2b0f4d/apps/editing-toolkit/editing-toolkit-plugin/common/index.php#L165 — I would be inclined to leave that as-is, and so that we're not adding extra complexity to the `ptk/patterns` endpoint until it's needed.
* Write up a feature request / idea or proposal for registering patterns via JS instead of PHP — I'm happy to take this task with me, so doesn't need to be tracked here.

I'll close this issue for now, but if any of the outstanding issues are of higher priority, it's probably worth opening individual issues for each of them 🙂
